File: 15xx/1557-minimum-number-of-vertices-to-reach-all-nodes/1557.c
/* 1557. Minimum Number of Vertices to Reach All Nodes */

/* Complexities.
 * Time: O(n + e)
 * Space: O(n)
 * where n is the number of vertices in the graph, and e its number of edges. */

#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

int *findSmallestSetOfVertices(int n, int **edges, int edgesSize,
                               int *edgesColSize, int *returnSize)
{
	/* booleans that state which vertices are sources */
	size_t space = (size_t) n * sizeof(int);
	int *isSource = malloc(space);
	memset(isSource, true, space);
	for (int i = 0; i < edgesSize; i++) {
		isSource[edges[i][1]] = false;
	}

	/* Gather together those vertices that are sources */
	int size = 0;
	int *tmp_result = malloc(space);
	int j = 0;
	for (int i = 0; i < n; ++i)
		if (isSource[i]) {
			tmp_result[j] = i;
			j++;
			size++;
		}
	/* the values in source are no longer needed */
	free(isSource);

	*returnSize = size;
	return reallocarray(tmp_result, (size_t) size, sizeof(int));
}

File: 15xx/1557-minimum-number-of-vertices-to-reach-all-nodes/1557.ts
/* 1557. Minimum Number of Vertices to Reach All Nodes */

/** The result are precisely those vertices that don’t have an incoming edge. **/

function findSmallestSetOfVertices(n: number, edges: number[][]): number[] {
  const source = new Array<boolean>(n);
  source.fill(true);
  for (let i = 0; i < edges.length; i++) {
    source[edges[i][1]] = false;
  }

  const result: number[] = [];
  for (let i = 0; i < n; i++) {
    if (source[i]) {
      result.push(i);
    }
  }

  return result;
}
